tests/test_typing.py

```diff
@@ -0,0 +1,158 @@
+from typing import Any, Optional
+
+from pytest import mark, param, raises
+
+from omegaconf import DictConfig, ListConfig, OmegaConf
+from omegaconf._utils import _ensure_container
+from omegaconf.errors import ValidationError
+from tests import (
+    ConcretePlugin,
+    Group,
+    OptionalUsers,
+    SubscriptedDict,
+    SubscriptedList,
+    User,
+    Users,
+)
+
+
+@mark.parametrize(
+    "cls,key,assignment,error",
+    [
+        param(SubscriptedList, "list", [None], True, id="list_elt"),
+        param(SubscriptedList, "list", [0, 1, None], True, id="list_elt_partial"),
+        param(SubscriptedDict, "dict_str", {"key": None}, True, id="dict_elt"),
+        param(
+            SubscriptedDict,
+            "dict_str",
+            {"key_valid": 123, "key_invalid": None},
+            True,
+            id="dict_elt_partial",
+        ),
+        param(SubscriptedList, "list", None, True, id="list"),
+        param(SubscriptedDict, "dict_str", None, True, id="dict"),
+        param(SubscriptedList, "opt_list", [None], True, id="opt_list_elt"),
+        param(SubscriptedDict, "opt_dict_str", {"key": None}, True, id="opt_dict_elt"),
+        param(SubscriptedList, "opt_list", None, False, id="opt_list"),
+        param(SubscriptedDict, "opt_dict_str", None, False, id="opt_dict"),
+        param(SubscriptedList, "list_opt", [None], False, id="list_opt_elt"),
+        param(SubscriptedDict, "dict_opt", {"key": None}, False, id="dict_opt_elt"),
+        param(SubscriptedList, "list_opt", None, True, id="list_opt"),
+        param(SubscriptedDict, "dict_opt", None, True, id="dict_opt"),
+        param(
+            ListConfig([None], element_type=Optional[User]),
+            0,
+            User("Bond", 7),
+            False,
+            id="Optional[User]<-User_succeeds",
+        ),
+        param(
+            ListConfig([User], element_type=User),
+            0,
+            None,
+            True,
+            id="Optional[User]<-str_fails",
```

Comment:
Two comments:
1. This kind of `id` formatting clashes with what's been used so far -- I don't really mind myself but just highlighting it for @omry 
2. In this one, should `str` be `None`?

Reply:
Right, thanks. I'd been meaning to change those (especially after our previous discussion on `id` conventions).

Updated in https://github.com/omry/omegaconf/pull/749/commits/a661d91000f87a157cf2355203b95443a462375e:
`"Optional[User]<-User_succeeds"` -> `"set_optional_user"`
`"Optional[User]<-str_fails"` -> `"illegal_set_user_to_none"`